### PR TITLE
wasm2js: Declare data segments before calling asmFunc

### PR DESCRIPTION
**Pull request by sbc100**
This is because we maybe need to reference the segments
during the start function.  For example in the case of
pthreads we conditionally load passive segments during
start.

Tested in emscripten with: tests/runner.py wasm2js1

### CODE REVIEW

File: src/wasm2js.h [hunk 1]
```diff
@@ -92,6 +92,39 @@ bool isTableExported(Module& wasm) {
   return false;
 }
 
+bool hasActiveSegments(Module& wasm) {
+  // If there are any active segments, initActiveSegments needs access
```
Comment (kripken) [inline src/wasm2js.h:96]:
```suggestion
```

File: src/wasm2js.h [hunk 2]
```diff
@@ -92,6 +92,39 @@ bool isTableExported(Module& wasm) {
...
+  // to bufferView.
```
Comment (kripken) [inline src/wasm2js.h:97]:
```suggestion
```

File: src/wasm2js.h (hunk 1 above)
Comment (kripken) [reply src/wasm2js.h:96]: (comment is duplicated from below, and not needed here)

File: src/wasm2js.h [hunk 3]
```diff
@@ -512,11 +545,23 @@ void Wasm2JSBuilder::addBasics(Ref ast, Module* wasm) {
...
-    if ((!wasm->memory.segments.empty()) || wasm->features.hasBulkMemory()) {
+    if (needsBufferView(*wasm)) {
       ast->push_back(
         ValueBuilder::makeBinary(ValueBuilder::makeName("bufferView"),
                                  SET,
                                  ValueBuilder::makeName(HEAPU8)));
+
+    }
+    if (hasActiveSegments(*wasm)) {
+      // If we have at least on active segment call `initActiveSegments` now.
```
Comment (kripken) [inline src/wasm2js.h:556]:
```suggestion
      // If we have at least one active segment, call `initActiveSegments` now.
```

File: src/wasm2js.h [hunk 4]
```diff
@@ -2574,16 +2598,35 @@ void Wasm2JSGlue::emitPostES6() {
...
+    if (seg.isPassive) {
+      out << "  var memorySegments = {};\n";
+      break;
+    }
+  }
+
   out <<
-    R"(for (var base64ReverseLookup = new Uint8Array(123/*'z'+1*/), i = 25; i >= 0; --i) {
+    R"(  var base64ReverseLookup = new Uint8Array(123/*'z'+1*/);
+  for (var i = 25; i >= 0; --i) {
```
Comment (kripken) [inline src/wasm2js.h:2629]: The old way is more compact (a single `var`).

**Review by kripken — APPROVED** [review]

File: src/wasm2js.h (hunk 4 above)
Comment (sbc100) [reply src/wasm2js.h:2623]: But it was horrendous to read.. I makes it look like `base64ReverseLookup` is scoped the for loop which it most certainly is not.  I don't think its worth the 4 bytes... if we want to go after 4 byte saving here I think there way much lower hanging fruit.

Comment (sbc100) [reply src/wasm2js.h:2623]: If you feel strongly I'd rather find other vars i this file to collapse.

Comment (sbc100) [reply src/wasm2js.h:2623]: For example couldn't we collapse all of these:

```
 var Math_imul = Math.imul;                                                      
 var Math_fround = Math.fround;                                                  
 var Math_abs = Math.abs;                                                        
 var Math_clz32 = Math.clz32;                                                    
 var Math_min = Math.min;                                                        
 var Math_max = Math.max;                                                        
 var Math_floor = Math.floor;                                                    
 var Math_ceil = Math.ceil;                                                      
 var Math_trunc = Math.trunc;                                                    
 var Math_sqrt = Math.sqrt;                                                      
 var abort = env.abort;   
```

?

Comment (kripken) [reply src/wasm2js.h:2621]: Fair enough, yeah, I don't feel strongly.

We should probably leave such micro-opts for closure anyhow. (which I think it does in closure=2)

**Pull request merged**